### PR TITLE
feat: update create/edit project forms to handle projects without credit class

### PR DESCRIPTION
**Pull request by blushi**
## Description

Closes: #2161

For projects without credit class:
- Removes the metadata form
- At the end of the create flow, do not submit the project on chain but just set it as "published" off-chain
- Do not submit any on-chain data when editing published off chain projects but keep storing everything off-chain

Also fixes: #2177 a bug that I came across while working on this

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] provided a link to the relevant issue or specification
- [x] provided instructions on how to test
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### How to test

Go to https://deploy-preview-2178--regen-marketplace.netlify.app/ and login with a class issuer/project admin.
From https://deploy-preview-2178--regen-marketplace.netlify.app/profile/projects:
1. Create a new project with a credit class and make sure the entire flow still works as before, in particular an on chain project is created at the end of the flow
1. Create a new project without any credit class:
   - no metadata form
   - the project shouldn't be created on chain but after final submission, it should be "published" meaning that if you go to the project page before this step, it will return 404 whereas if you go to the project page after the final review step submission, it will show the project data (use the UUID in the URL to go to the project page: https://deploy-preview-2178--regen-marketplace.netlify.app/project/{someUUID})
1. Edit an on chain project, behavior should be unchanged: on chain actions gets triggered when editing anchored data (basic info, location, roles and metadata)
1. Edit a project without a credit class that hasn't been published or a project with a credit class that is not on chain yet, you should still be redirected to the project create flow
1. Edit a project without any credit class that has been published:
   - no metadata form
   - you should be redirected to the project edit flow and editing shouldn't trigger any on chain actions (contrary to 3.)

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please **add
your handle next to the items reviewed if you only reviewed selected items**._

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed code correctness and readability
- [ ] verified React components follow DRY principles
- [ ] reviewed documentation is accurate
- [ ] reviewed tests
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by blushi** [comment]
@erikalogie see testing instructions above

**Comment by erikalogie** [comment]
@blushi when I try to hit save & next on the review step for the offchain project creation, nothing happens. I'm getting these errors in the console:
Screen Shot 2023-10-12 at 10 02 36 AM

This is "Offchain project test"  that I'm having trouble with.

**Comment by erikalogie** [comment]
Also I don't see the "Offchain project test" in my list of projects. Shouldn't it be saved there before it is "published"?

**Comment by blushi** [comment]
> @blushi when I try to hit save & next on the review step for the offchain project creation, nothing happens. I'm getting these errors in the console: Screen Shot 2023-10-12 at 10 02 36 AM
> 
> This is "Offchain project test" that I'm having trouble with.

These logs don't seem to be related (coming to wallet connect which is not enabled on this deploy preview), when clickcing "save & next", could you go to the "network" tab instead and if you see a red line there, click and share the "Payload" and "Preview" tabs, thanks.

It looks like this might be related to caching issue, I'll look into it. If you reload the profile/projects page, you should see "Offchain project test".

**Comment by blushi** [comment]
> These logs don't seem to be related (coming to wallet connect which is not enabled on this deploy preview), when clicking "save & next", could you go to the "network" tab instead and if you see a red line there, click and share the "Payload" and "Preview" tabs, thanks.

Ok nevermind, I've found the issue, this was related to some database permission, which I've manually fixed for now on the staging db, this is handled in @wgwz's PR which is not merged yet: https://github.com/regen-network/regen-server/pull/398/commits/9dad660f3c6451bee7ed8f8a9ab5f3106157a926

File: web-marketplace/src/pages/Roles/hooks/useRolesSubmit.tsx [hunk 1]
```diff
@@ -117,8 +123,10 @@ const useRolesSubmit = ({
         // In creation or edit mode, we always store references to the project stakeholders in the project table
         // which should be in projectPatch if new or updated
         if (doUpdateMetadata || doUpdateAdmin) {
-          // In edit mode, we need to update the project on-chain metadata and/or admin if needed
-          if (isEdit) {
+          console.log(onChainProject);
```
Comment (flagrede) [inline web-marketplace/src/pages/Roles/hooks/useRolesSubmit.tsx:126]: to remove?

**Review by flagrede — APPROVED** [review]
tACK

Nice work on this one. I tested both flows (existing and new ones) and was able to perform all actions as expected.

**Comment by blushi** [comment]
@erikalogie you can test this again, thanks.

**Comment by erikalogie** [comment]
When I go to edit my offchain project, I get taken to the review step of the project creation flow instead of the edit form.


Uploading Screen Recording 2023-10-18 at 11.20.18 AM.mov…

**Comment by blushi** [comment]
> When I go to edit my offchain project, I get taken to the review step of the project creation flow instead of the edit form.
> 
> Uploading Screen Recording 2023-10-18 at 11.20.18 AM.mov…

looks like your video has been uploaded

**Comment by blushi** [comment]
> > When I go to edit my offchain project, I get taken to the review step of the project creation flow instead of the edit form.
 
it could be because the project hasn't been published yet (ie you didn't hit the final save button in the review step), see item 4. of the testing instructions above
if you did that, that might be a bug indeed
could you share the project uuid?

**Comment by erikalogie** [comment]
> > When I go to edit my offchain project, I get taken to the review step of the project creation flow instead of the edit form.
> 
> it could be because the project hasn't been published yet (ie you didn't hit the final save button in the review step), see item 4. of the testing instructions above if you did that, that might be a bug indeed could you share the project uuid?

Yeah I did save it and got to the final publish confirmation screen.

**Comment by erikalogie** [comment]
I can't get the UUID because the project has now disappeared from the projects lists, I'm so confused

**Comment by blushi** [comment]
> I can't get the UUID because the project has now disappeared from the projects lists, I'm so confused

I can see it in your video so will investigate

**Comment by erikalogie** [comment]
ok wait I reloaded and now it is there again: 7adbbfc2-6dc9-11ee-b14a-0267c2be097b

**Comment by erikalogie** [comment]
Now when I click the edit button it goes to the correct place, to the edit form

**Comment by erikalogie** [comment]
> https://deploy-preview-2178--regen-marketplace.netlify.app/

@blushi this is what I get: 

Screen Shot 2023-10-18 at 12 36 56 PM

Screen Shot 2023-10-18 at 12 36 51 PM

**Comment by blushi** [comment]
> > https://deploy-preview-2178--regen-marketplace.netlify.app/
> 
> @blushi this is what I get:
> 
> Screen Shot 2023-10-18 at 12 36 56 PM Screen Shot 2023-10-18 at 12 36 51 PM

could you share the "payload" tab content too? thanks

**Comment by erikalogie** [comment]
Ok, it weirdly isn't happening for me anymore now

**Comment by blushi** [comment]
> Ok, it weirdly isn't happening for me anymore now

what exactly?

**Comment by erikalogie** [comment]
> > Ok, it weirdly isn't happening for me anymore now
> 
> what exactly?

Both things, now clicking on the edit button for a newly created off chain project takes me to the edit form for that project and I no longer get the red banner and error creating project error when trying to create a new project after offchain project creation.

**Comment by blushi** [comment]
> I no longer get the red banner and error creating project error when trying to create a new project after offchain project creation.

Given my recent findings, as soon as you reload the page, this should work again yeah (this is a weird edge case that happens as soon as you have multiple tabs/windows open with the app so unrelated to this, it will be fixed separately)

**Comment by blushi** [comment]
> > Both things, now clicking on the edit button for a newly created off chain project takes me to the edit form for that project
> 
> When you say "newly created off chain project", do you mean it it got published too (ie clicking save in the review process)? As I said the issue that you described seemed related to data caching so as soon as the page gets reloaded, it should work as expected with the correct data.
> 
> > I no longer get the red banner and error creating project error when trying to create a new project after offchain project creation.
> 
> Given my recent findings, as soon as you reload the page, this should work again yeah (this is a weird edge case that happens as soon as you have multiple tabs/windows open with the app so unrelated to this, it will be fixed separately)

@erikalogie up
Have you been able to test this again? I assume some of the issues you were having are due to the CSRF token issue, which happens as soon as you have multiple tabs/windows open with the app, as I explained in https://github.com/regen-network/regen-web/issues/2085#issuecomment-1770436847. Could you test again with just one tab open? Thanks

**Comment by erikalogie** [comment]
> > > Both things, now clicking on the edit button for a newly created off chain project takes me to the edit form for that project
> > 
> > 
> > When you say "newly created off chain project", do you mean it it got published too (ie clicking save in the review process)? As I said the issue that you described seemed related to data caching so as soon as the page gets reloaded, it should work as expected with the correct data.
> > > I no longer get the red banner and error creating project error when trying to create a new project after offchain project creation.
> > 
> > 
> > Given my recent findings, as soon as you reload the page, this should work again yeah (this is a weird edge case that happens as soon as you have multiple tabs/windows open with the app so unrelated to this, it will be fixed separately)
> 
> @erikalogie up Have you been able to test this again? I assume some of the issues you were having are due to the CSRF token issue, which happens as soon as you have multiple tabs/windows open with the app, as I explained in [#2085 (comment)](https://github.com/regen-network/regen-web/issues/2085#issuecomment-1770436847). Could you test again with just one tab open? Thanks

Yes sorry it is working as expected.
